src/bin/pg_autoctl/debian.c

```diff
@@ -0,0 +1,699 @@
+/*
+ * src/bin/pg_autoctl/debian.c
+ *
+ *   Debian specific code to support registering a pg_autoctl node from a
+ *   Postgres cluster created with pg_createcluster. We need to move the
+ *   configuration files back to PGDATA.
+ *
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the PostgreSQL License.
+ *
+ */
+#include <libgen.h>
+
+#include "postgres_fe.h"
+#include "pqexpbuffer.h"
+
+#include "debian.h"
+#include "keeper.h"
+#include "keeper_config.h"
+#include "parsing.h"
+
+#define EDITED_BY_PG_AUTOCTL "# edited by pg_auto_failover \n"
+
+static bool debian_find_postgres_configuration_files(
+	PostgresSetup *pgSetup,
+	PostgresConfigFiles *pgConfigFiles);
+
+static bool debian_init_postgres_config_files(
+	PostgresSetup *pgSetup,
+	PostgresConfigFiles *pgConfFiles,
+	PostgresConfigurationKind confKind);
+
+static bool buildDebianDataAndConfDirectoryNames(
+	PostgresSetup *pgSetup,
+	DebianPathnames *debPathnames);
+
+static bool expandDebianPatterns(
+	DebianPathnames *debPathnames,
+	const char *dataDirectoryTemplate,
+	const char *confDirectoryTemplate);
+
+static bool expandDebianPatternsInDirectoryName(
+	char *pathname,
+	int pathnameSize,
+	const char *template,
+	const char *versionName,
+	const char *clusterName);
+
+static void initPostgresConfigFiles(
+	const char *dirname,
+	PostgresConfigFiles *pgConfigFiles,
+	PostgresConfigurationKind kind);
+
+static bool postgresConfigFilesAllExist(
+	PostgresConfigFiles *pgConfigFiles);
+
+static bool move_configuration_files(
+	PostgresConfigFiles *src,
+	PostgresConfigFiles *dst);
+
+static bool comment_out_configuration_parameters(
+	const char *srcConfPath,
+	const char *dstConfPath);
+
+static bool disableAutoStart(PostgresConfigFiles *pgConfigFiles);
+
+
+/*
+ * keeper_ensure_pg_configuration_files_in_pgdata checks if postgresql.conf,
+ * pg_hba.conf, pg_ident.conf files exist in $PGDATA, if not it tries to get
+ * them from default location and modifies paths inside copied postgresql.conf.
+ */
+bool
+keeper_ensure_pg_configuration_files_in_pgdata(KeeperConfig *config)
+{
+	PostgresSetup *pgSetup = &(config->pgSetup);
+	PostgresConfigFiles pgConfigFiles = { 0 };
+
+	if (!debian_find_postgres_configuration_files(pgSetup, &pgConfigFiles))
+	{
+		/* errors have already been logged */
+		return false;
+	}
+
+	switch (pgConfigFiles.kind)
+	{
+		case PG_CONFIG_TYPE_POSTGRES:
```

Comment:
it is always of type POSTGRES,  could not think of a better name though. DEFAULT ? but we have way too many "default"s around

Reply:
I meant Postgres as in “like the Postgres core code” to differentiate with “like the debian patched version” or nowadays “like the debian packaging”... maybe I should add more comments about that?

Reply:
Actually debian.h contains the following, I think that's all we need?

```
 * We know how to find configuration files in either PGDATA as per Postgres
 * core, or in the debian cluster configuration directory as per debian
 * postgres-common packaging, implemented in pg_createcluster.
```